src/transformers/models/albert/modeling_albert.py

```diff
@@ -887,8 +887,12 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
         return prediction_scores
 
     def _tie_weights(self) -> None:
-        # To tie those two weights if they get disconnected (on TPU or when the bias is resized)
-        self.bias = self.decoder.bias
+        # For accelerate compatibility and to not break backward compatibility
+        if self.decoder.bias.device.type == "meta":
+            self.decoder.bias = self.bias
+        else:
+            # To tie those two weights if they get disconnected (on TPU or when the bias is resized)
+            self.bias = self.decoder.bias
```

Comment:
For my own understanding could you provide a bit more information here about the two cases. In particular, my questions are:
* Which of the two branches of if/else here are for accelerate compatibility?
* How was accelerate compatibility caught i.e. which tests? 
* Do we need both of these assignments? I'm assuming yes but it's not clear to me why one works in one case and not the other.

Reply:
I copied this from existing tie_weights():
https://github.com/huggingface/transformers/blob/f3aa7db439a2a3942f76c115197fe953984ac334/src/transformers/models/roberta/modeling_roberta.py#L1141

I agree that the existing implementation of `tie_weights()` doesn't make sense. However, I added the if-statement to keep backward compatibility. 

Here's my 2 cents on how this all works:

When loading the model, only one pointer out of the tied params would be loaded with the correct values (when device=meta). Let's call this the "canonical" pointer. All the other tied params must copy from the canonical pointer, and it can't really be the other way around (at least for device=meta case).

The canonical pointer is the key stored in the loaded `state_dict`. In the current logic for [`save_pretrained()`](https://github.com/huggingface/transformers/blob/f3aa7db439a2a3942f76c115197fe953984ac334/src/transformers/modeling_utils.py#L2395), the "canonical" pointer is the weight key that's not in `_tied_weight_keys` list. In this case, that would make `self.bias` the canonical pointer.

However, I imagine it's still possible for some (older?) pretrained model to have some other logic of choosing which is the "canonical" pointer, and so I added the if-statement for backwards compatibility.

Reply:
One note about the case where device != meta. When that happens, the direction of the pointer assignment doesn't matter as much. Whether you call `self.bias = self.decoder.bias` or `self.decoder.bias = self.bias`, when the params are loaded, they show up in both since they almost act like a pointer reference.

When device == meta, the pointer assignment operator doesn't have the same pointer reference relationship. It just copies the sizes/info of the other tensor. When one of those tensors are loaded (the canonical one), the other one needs to copy the pointer reference accordingly, it is not automatically loaded the same way.